### PR TITLE
Added Const and Iterate

### PR DESCRIPTION
**Pull request by Bolpat**
In an informal nutshell: (`*s` can be a sequence)
- `(Const!Xs)!Ys == Xs;` where `Ys` is not evaluated.
- `Iterate!(F, n, Xs) == F!(F!(...F!(Xs)...))` with `n` applications of `F`

Not in this PR anymore:
- <s>Iota</s>
- <s>Replicate</s>

These functions have been requested once in a while in the learning forum. Are there any cons against having them in the library?
If someone has ideas to further generalize them, please tell me.

### CODE REVIEW

**Comment by BBasile** [comment]
There are two problems with `Iota`:
- it's already defined in `std.experimental.ndslice` internal package (you've probably noticed the auto-tester failure)
- phobos __already__ defines something similar: `std.typecons.staticIota`, but its protection is `package`.

Considering the last point, I think that it would be better to
- remove `Iota` from this PR
- in a new PR
  - rename your template to `staticIota`, add this template to `std.meta`.
  - alias the one from typecons to this new version (since yours is a bit better).

There's already been discussion about the existing `staticIota`, saying that it should be public. Now that the bug 314 is fixed, I think it's a good opportunity to replace it and make it public.

**Comment by JakobOvrum** [comment]
std.meta.aliasSeqOf superseded staticIota:
```d
static assert(aliasSeqOf!(iota(1, 6, 2)) == AliasSeq!(1, 3, 5));
```

**Comment by e10s** [comment]
`Replicate` looks quite similar to `std.mete.Repeat`, though the latter does not allow `n==0`. In fact, it was written by me and the design is based on the idea that no repetition would be meaningless.
If necessary, it can be modified for zero `n`.

File: std/meta.d [hunk 1]
```diff
@@ -169,6 +169,128 @@ unittest
...
+    static assert(is(X!uint == int));
+
+    alias Y = Const!(uint, uint);
+    static assert(is(Y!(int, uint) == AliasSeq!(uint, uint)));
+}
+
+/**
+ * Creates a sequence values, similar to how std.range.iota does, but as a sequence not a range.
+ */
+template Iota(int stop)
```
Comment (9il) [inline std/meta.d]: Could you please replace type with `sizediff_t`?

**Comment by Bolpat** [comment]
@BBasile Your suggestions are sensible, so I just removed Iota. If the workaround is that easy like @JakobOvrum stated, maybe we just don't need `Iota` at all. As this is for a new pull request, let's discuss that there.
@e10s Somehow I must have overseen `Repeat` (looking embarrassed to the ground...). I think making `Repeat` allowing `n = 0` is easy, sometimes useful and nobody will have serious problems with it. In contrary, sooner or later someone will expect it to work for `n = 0` in this way.
@9il I've never seen `sizediff_t`. It seems to be very unknown and Google has no results searching the language spec. Is it the same as `ptrdiff_t`? I've changed it to that. If there is a reason for using `sizediff_t`, I'll just change it again.

**Comment by BBasile** [comment]
I'm sorry but it seems, according to Jakob O, that I was wrong. I missed that `aliasSeqOf` was added to phobos and my comment is actually based on an older discussion related to a Q. on stackoverflow.

File: std/meta.d [hunk 2]
```diff
@@ -1177,8 +1228,12 @@ unittest
...
  * Creates an `AliasSeq` which repeats a type or an `AliasSeq` exactly `n` times.
  */
-template Repeat(size_t n, TList...) if (n > 0)
+template Repeat(size_t n, TList...)
 {
+    static if (n == 0)
+    {
+        alias Repeat = AliasSeq!();
+    }
     static if (n == 1)
```
Comment (e10s) [inline std/meta.d]: `else static if (n == 1)`

File: std/meta.d [hunk 3]
```diff
@@ -1177,9 +1228,13 @@ unittest
...
 {
-    static if (n == 1)
+    static if (n == 0)
+    {
+        alias Repeat = AliasSeq!();
+    }
+    else static if (n == 1)
     {
         alias Repeat = AliasSeq!TList;
     }
```
Comment (e10s) [inline std/meta.d]: `else static if (n == 2){...}` is no longer necessary.
And, could you please add a test for `Repeat!(0, X)`?

Comment (Bolpat) [inline std/meta.d]: Doesn't `n == 2` speed up compilation?

Comment (e10s) [inline std/meta.d]: That's a good point.

**Comment by Bolpat** [comment]
Sorry, this was a mistake.

File: std/meta.d [hunk 4]
```diff
@@ -169,6 +169,57 @@ unittest
...
+ * Returns $(D F!(F!(...(F!(X))...))) with n iterations of F.
+ */
+template Iterate(alias F, size_t n, X...)
+{
+    static if (n == 0)
+    {
+        static if (X.length == 1)
+            alias Iterate = X[0];
+        else
+            alias Iterate = X;
```
Comment (DmitryOlshansky) [inline std/meta.d]: Wouldn't Iterate = X just work?

File: std/meta.d [hunk 5]
```diff
@@ -169,6 +169,57 @@ unittest
...
+    static assert(is(X!uint == int));
+
+    alias Y = Const!(uint, uint);
+    static assert(is(Y!(int, uint) == AliasSeq!(uint, uint)));
+}
+
+/**
+ * Returns $(D F!(F!(...(F!(X))...))) with n iterations of F.
+ */
+template Iterate(alias F, size_t n, X...)
```
Comment (DmitryOlshansky) [inline std/meta.d]: This is certainly not quite the iteration but rather application performed  n times.

Comment (DmitryOlshansky) [inline std/meta.d]: So it's more close to power in mathematical sense

File: std/meta.d [hunk 6]
```diff
@@ -1177,9 +1228,13 @@ unittest
 /**
  * Creates an `AliasSeq` which repeats a type or an `AliasSeq` exactly `n` times.
  */
-template Repeat(size_t n, TList...) if (n > 0)
+template Repeat(size_t n, TList...)
 {
-    static if (n == 1)
+    static if (n == 0)
+    {
+        alias Repeat = AliasSeq!();
```
Comment (DmitryOlshansky) [inline std/meta.d]: nice

File: std/meta.d (hunk 5 above)
Comment (e10s) [inline std/meta.d]: Which is catchier or more appropriate?
- [iterated function](https://en.wikipedia.org/wiki/Iterated_function)
- [functional power](https://en.wikipedia.org/wiki/Function_composition#Functional_powers)

**Comment by Bolpat** [comment]
I'm closing this and open a new PR with `Iterate` only. So far, nobody has complained about it. Then we can discuss if it is worth having it in Phobos.

**Pull request closed without merging**